File: 007.c
#include <stdio.h>
#include <math.h>

// Ejercicio Nº 5 de la guía Nº 2.

/**
 * Función Cos(x) - x
 * 
 * @author Daniel Cantarín
 * @date 20160421
 **/
double f1( double );


/**
 * Función f1'
 * 
 * @author Daniel Cantarín
 * @date 20160421
 **/
double fd1( double );


int main() 
{
    double x0, x1, cota, error;
    printf("Ingrese x0:");
    scanf("%lf", &x0);
    
    printf("Ingrese margen de error:");
    scanf("%lf", &error);
    
    cota = 50;
    while ( cota > error ) { 
        x1   = x0 -  f1(x0) / fd1(x0);
        cota = fabs(x1 - x0);
        //printf("Error: %lf, Cota: %lf, x0: %lf, x1: %lf\n", error, cota, x0, x1);
        x0   = x1;
    }
    
    printf("Tomá: %lf\n", x0);
    
    return 0;
}


double f1 ( double x ) 
{
    double z;
    
    z = x - cos(x);
    
    return z;
}


double fd1 ( double x ) 
{
    double z;
    
    z = 1 + sin(x);
    
    return z;
}

File: 007.m
# Derivada y otras funciones.
# @author   Daniel Cantarín <canta@canta.com.ar>
# @date     20160908

function salida = secante_aux( f, a, b )
    
    f0 = feval( f , a);
    f1 = feval( f , b);
    
    c  = (a - b) + 0.001;
    
    salida = (f0 - f1) / c;
endfunction


function salida = secante( f, a, b , cota )
    
    i      = 1;
    x0     = a;
    ret(i) = b;
    comp   = abs(a - b);
    
    while comp > cota
        i      = i + 1;
        f0     = feval( f , a);
        f1     = feval( "secante_aux" , f, a, b) + 0.001;
        b      = a - ( f0 / f1 );
        comp   = abs(a - b);
        a      = b;
        if (b == 0)
            comp = cota;
        end
        ret(i) = b;
    end
    
    salida = ret;
endfunction


function salida = newton( x0, f, f1, cota )
    
    x1     = 0;
    i      = 1;
    ret(i) = x0;
    comp   = cota  * 100;
    
    while comp > cota
        i       = i + 1;
        x1      = x0 - ( feval(f, x0) / feval( f1, x0 ) );
        comp    = abs(x0 - x1);
        x0      = x1;
        ret(i)  = x1;
    end
    
    salida = ret;
    
endfunction

function salida = menos_seno( a )
    
    salida = sin( a ) * (-1);
    
endfunction
